### DISCUSSION TITLE
Help switching from `getStorage`, `serialize` and `deserialize` to `storage`

### DISCUSSION THREAD

**Discussion by stephanedupont** [pmndrs/zustand#1584] [Q&A] [answered]
Hello,

I just updated Zustand and have some deprecated warning in the console. I looked at the documentation that mention this, but I still have trouble understand how to upgrade to `storage`.

Here's my current code:

```
const VERSION = 1;

export const cetAppDataStore =
  createStore(
    persist(
      redux(getReducer("CET Data"), appDataStoreBlankState()),
      {
        name: "cetAppData",
        version: VERSION,
        getStorage: () => localStorage,
        serialize: (state) =>
          StoreUtils.serializeStore(VERSION, state, CONFIG_ENCRYPT_STORES),
        deserialize: (str) => StoreUtils.deserializeStore(str, CONFIG_ENCRYPT_STORES),
      }
    )
  );
```

... with the following helper methods:

```
  static serializeStore(version: number, state: LooseObject, encrypt: boolean) {
    let storeState;
    if (encrypt) {
      storeState = {
        version: version,
        state: AppLowLevelUtils.encrypt(JSON.stringify(state.state))
      };
    } else {
      storeState = state;
    }
    return JSON.stringify(storeState)
  }

  static deserializeStore(str: string, encrypted: boolean) {
    const parsed = JSON.parse(str);
    if (encrypted) {
      parsed.state = JSON.parse(AppLowLevelUtils.decrypt(parsed.state));
    }
    return parsed;
  }
```

Any help would be greatly appreciated!

**Comment by dai-shi** [comment, marked as answer]
Thanks for asking. If we resolve this, you could help improving docs!

So, the storage type is the following:
https://github.com/pmndrs/zustand/blob/ab4118d395e3cd65a26eb53b639ac0616b270c51/src/middleware/persist.ts#L14-L20

Your custom storage will be something like this:
```js
const storage = {
  getItem: (name) => StoreUtils.deserializeStore(localStorage.getItm(name), CONFIG_ENCRYPT_STORES),
  setItem: (name, value) => localStorage.setItm(name, StoreUtils.serializeStore(value, CONFIG_ENCRYPT_STORES)),
  removeItem: (name) => localStorage.removeItem(name),
}
```
You need to change your util a little bit:
```js
  static serializeStore(value: { version: number; state: LooseObject }, encrypt: boolean) {
    let storeState;
    if (encrypt) {
      storeState = {
        version: value.version,
        state: AppLowLevelUtils.encrypt(JSON.stringify(value.state))
      };
    } else {
      storeState = value;
    }
    return JSON.stringify(storeState)
  }
```

**Comment by stephanedupont** [comment]
Thank you so much for the reply. The quality of support for this library is incredible!

This works great, and I have a couple questions left:

1. It seems that `value.version` is not always set in `setItem: (name, value) => ...`, I have the followin TS error:

```
TS2345: Argument of type 'StorageValue<Write<DataStoreState, ReduxState<DataStoreAction>>>' is not assignable to parameter of type '{ version: number; state: LooseObject; }'.   Types of property 'version' are incompatible.     Type 'number | undefined' is not assignable to type 'number'.       Type 'undefined' is not assignable to type 'number'.
```

Are there some cases where the version is null/undefined, or is it a TS def problem in Zustand?

2. For a simpler case where I didn't want to encrypt the store, I had the simpler following code:

```
const VERSION = 1;
const NAME = "data";

export const useSettingsStore =
  create(
    persist(
      redux(reducer, initialState),
      {
        name: NAME,
        version: VERSION,
        getStorage: () => localStorage,
      }
    )
  );
```

To migrate this deprecated usage, do I need to add the three methods, something like this:

```
storage: {
    getItem: (name) => localStorage.getItem(name),
    setItem: (name, value) => localStorage.setItem(name, value),
    removeItem: (name) => localStorage.removeItem(name),
},
```
... or can I simplify this for this simple case?

(Also, the code above triggers a TS warning about `string | null` not being assignable to what `getItem` expects, and about `value` not being of the `string` type.)

Once solved, shall I then add this example of "migration" in the docs?

Thank you very much!

**Comment by stephanedupont** [comment]
OK, for the 1., I added a defaultVersion param to the method to set it if not defined, and for the 2., I realized that to use localStorage, I could just not specify anything.

Thank you for the help!

**Reply by dai-shi** [reply]
Sorry for the delay. So, what's the remaining question? Thanks.

**Reply by stephanedupont** [reply]
None :)
